### PR TITLE
Openstack modules

### PR DESCRIPTION
**Pull request by nlewo**
I continued the works started by @domenkozar on Openstack NixOS modules.
I'm now able to deploy an Openstack compute node, create a network, a floating IP, boot a VM and ssh to it. So all basic components seem to work well;)
Note the Neutron driver is currently linuxbridge but I plan to add OVS.

There is still some (a lot of?) work to have clean modules. 
I would like to know if some people are interested in these modules and if we could define what is required to go to a merge in master.

For those who are not familiar with Openstack but want to try it, I wrote a short [guide](http://lewo.abesis.fr/posts/2016-09-13-openstack-on-nixos.html) to easily play with these modules, thanks to nixops.

### CODE REVIEW

**Comment by domenkozar** [comment]
Cool :) I've used this work for https://github.com/domenkozar/snabb-openstack-testing and never got to finishing the cleanup phase, which needs quite some work as you said :)

**Comment by nlewo** [comment]
- The test has been updated to ping a VMs from the hypervisor through the public network thanks to the neutron linuxbridge driver.
- All credentials and token are variabilized.
- Admin and service tenants are created in the systemd post start section.

@domenkozar, do you know someone who could be interested to comment or review this PR ?

**Comment by badi** [comment]
@nlewo Thank you for this. I saw your post on the mailing list and was eager to try it.
I had to modify `openstack.nix` from your [guide](http://lewo.abesis.fr/posts/2016-09-13-openstack-on-nixos.html). Namely:

```
virtualisation.keystone.enableSingleNode
```

does not exist in 59daf08

I had to instead set

```
virtualisation.keystone.enable
```

I'm also deploying on to virtualbox rather than libvirtd.

---
Below is the full `openstack.nix` I'm using.
Deploying with 

```
$ git rev-parse HEAD
59daf08b6fa388828484e4f01f62c15aa73660a0
$ nixops deploy -d openstack -I nixpkgs=`pwd`
```

```nix
{

  openstack =
    {config, pkgs, nixpkgs, ...}:
    let

      credentials = {
        keystoneAdminUsername = "admin";
        keystoneAdminTenant = "admin";
        keystoneAdminPassword = "admin";
      };
      osConfig = {
        endpointPublic = config.networking.privateIPv4;
      } // credentials;
      keystoneConfig = {
        enable = true;
      } // osConfig;
      otherConfig = {
        enableSingleNode = true;
      } // osConfig;

    in

    {

      deployment.targetEnv = "virtualbox";
      deployment.virtualbox.memorySize = 4096;
      deployment.virtualbox.headless = true;

      nixpkgs.config.allowBroken = true;

      networking.firewall.enable = false;

      environment.systemPackages = with pkgs.pythonPackages;
        [ neutronclient
          keystoneclient
          glanceclient
          novaclient
        ];

      virtualisation.keystone = keystoneConfig;
      virtualisation.glance = otherConfig;
      virtualisation.neutron = otherConfig;
      virtualisation.nova = otherConfig;

    };

}
```

File: nixos/modules/virtualisation/neutron.nix
Comment (spinus) [inline nixos/modules/virtualisation/neutron.nix]: Here you could do a nice trick. Becasue keystone can hang for whatever reason, you could prefix it with timeout i.e: "timeout 30 keystone...."

Next thing (just to consider) is to use $SECONDS variable instead of count, so you know for how long you will be trying. ($SECONDS gives you seconds since bash started, so you could check condition like : 
```
max=$(($SECONDS+30))
if [ $SECONDS < $max ];then
  break
fi
```
)

**Review by spinus — CHANGES_REQUESTED** [review]
Hey, I did small review. Biggest thing I noticed is that there is a lot of passwords in the configs and scripts which probably could be extracted to files (so for example files with secrets can be deployed by "nixops" or other secret managment, instead of putting them in nix store.

I'm new to open stack but I'm interested in that work and will be happy to review test and support with commit power if needed. Please ping me about that whenever you want.

**Comment by nlewo** [comment]
@badi thx for your feedback. I upgraded the blog post according to your remarks. There is now a nixos test that builds a VM ready to use. I think this is the most convenient way to do some tries.

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: @spinus Could you provide a pointer related to secret managment? I don't know what is secret managment good pratice.

Comment (spinus) [inline nixos/modules/virtualisation/glance.nix]: @nlewo I don't think we have something like that. But there is known thing that all derivations end up in /nix/store so all the passwords go there.

What I do to avoid it is:
- I follow nixops pattern to put ALL credentials in the key store (https://nixos.org/nixops/manual/#opt-deployment.keys) - those keys are not going to be written in /nix/store, they will end up in /run/keys/<keyname>.
- in the scripts I use:
```
run-something --secret "$(</run/keys/<my-secret>)"
```
Config files are little harder. If it's possible to use that by the process, use mix of config files and environment variables. If it's not possible, I keep template of config file in nix store, and than, on preStart script I copy this file to /var/lib/<service>/ and fill the template (for example with sed) with keys from /run/keys.

File: nixos/lib/make-disk-image.nix
Comment (nlewo) [inline nixos/lib/make-disk-image.nix]: I don't know why this was set... so I remove it and tests are still ok.

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: In fact, we are currently only support single node. I think we could address multinode setup in future PR.
I update the description to be more explicit about this limitation

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: Done.

File: nixos/modules/virtualisation/neutron.nix
Comment (nlewo) [inline nixos/modules/virtualisation/neutron.nix]: This interface is added to the bridge created for an network which is 'bound' to an external one. I update the documentation to provide some help about this.

File: nixos/modules/virtualisation/neutron.nix
Comment (nlewo) [inline nixos/modules/virtualisation/neutron.nix]: Added `requires`.

File: nixos/modules/virtualisation/neutron.nix
Comment (nlewo) [inline nixos/modules/virtualisation/neutron.nix]: Done.

File: nixos/modules/virtualisation/neutron.nix
Comment (nlewo) [inline nixos/modules/virtualisation/neutron.nix]: I prefer to let it as it is since this is not really a timeout. We try 30 times but each try can be longer than 1 seconds.

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: I would prefer to not do this now since this PR is already really big. This will add more complexity. Moreover, current modules target a dev environment, where password are not really important.
What do you think to address this problem in another PR (once/if this one is merged)?

**Comment by nlewo** [comment]
@spinus thanks for all of your comments.

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: yep, I also think I have to create a PR per components.

**Comment by spinus** [comment]
@nlewo no worries. I'm interested in helping with this, I was just too lazy to start it myself but if you need a tester, code reviews or some other help let me know.

File: nixos/modules/virtualisation/glance.nix
Comment (ip1981) [inline nixos/modules/virtualisation/glance.nix]: It's the best when mysql client uses mysql client config files only. They allow `!include`, and thus you could have non-sensitive parameters in a "public" file and secret file included in it.

File: nixos/modules/virtualisation/keystone.nix
Comment (ip1981) [inline nixos/modules/virtualisation/keystone.nix]: This suggests that you use local mysql only. If so, you don't need any passwords: use UNIX-socket authentication. But I think local mysql is not a strategy, and then the service better not to depends on mysql, or poll it.

File: nixos/modules/virtualisation/glance.nix
Comment (nlewo) [inline nixos/modules/virtualisation/glance.nix]: @ip1981 Sorry, I don't understand your point. Could you point me a doc or an example related to the pattern you are suggesting?

Comment (ip1981) [inline nixos/modules/virtualisation/glance.nix]: Like this https://github.com/ip1981/mywatch#configuration

**Comment by nlewo** [comment]
I start to split this PR into several sub PR. The first one is keystone #20905.

**Comment by domenkozar** [comment]
Sorry as I don't have time to review this, but others have jumped on the ship. Thanks for getting this upstream!

Just a reference, here is WIP backend for nixops: https://github.com/NixOS/nixops/pull/408

**Comment by nlewo** [comment]
@domenkozar, this PR is too big and really not attractive for reviewers... It's why I submitted a PR that only addresses Keystone #20905. For sake of clarity, I squashed our commits into one and I mentioned your works in the commit message (I hope you're ok with that).

**Comment by spinus** [comment]
@domenkozar what do you think about passwords and secrets, should this be passed as file (or env variable, or other) or still worth to merge this with passwords in nix store as a first draft?

**Comment by ip1981** [comment]
> worth to merge this with passwords in nix store as a first draft

why rush?

**Comment by spinus** [comment]
@ip1981 no rush, I'm just exposing some comments
check :
https://github.com/NixOS/nixpkgs/pull/19529#discussion_r90426861
https://github.com/NixOS/nixpkgs/pull/19529#discussion_r90547298

**Comment by domenkozar** [comment]
@spinus I don't know how to handle that best. It's the reason I haven't submitted this upstream. They shouldn't be in store, but we do have some services that do it today, but eventually someone will be screwed.

**Comment by spinus** [comment]
@domenkozar that is my impression too hence I suggested to refactor this to keep secrets out of nix store but wanted some confirmation from the core team. Cheers.

**Comment by grahamc** [comment]
Hi @nlewo, I am thrilled that you're putting this effort in. I hope the (long) delay to getting this merged hasn't discouraged you too much.

One thing I've noticed is basically all our current openstack modules need upgrades. Is that a project you could undertake as well?

**Comment by nlewo** [comment]
Hi @grahamc, I understand it takes time to upstream all these modules and since this is the first time I'm doing this... it takes even more times!

I already planned to upgrade our openstack version, but I would prefer to first have modules for basic openstack modules in order to be able to easily do functional tests. 
Keystone is merged and I'll submit glance in the next few days.

Do you have an idea (order of magnitude) on how many time we could continue to have the liberty release?  (I know several production lines where Juno is still in use...)
This would allow me to know if I can continue to upstream nova and neutron modules or if I have to first upgrade openstack packages.

Are there known security issues on liberty components?

Thanks for your interest.

**Comment by Mic92** [comment]
Glance and Keystone are now already part of nixos.

**Comment by nlewo** [comment]
I will create new PRs regarding neutron and nova (as I did for keystone and glance). So I close this one mainly to clean the backlog;)

**Pull request closed without merging**
